diff --git a/src/module/AdiabaticCooling.cpp b/src/module/AdiabaticCooling.cpp
index 40ccb45..6a8a1ec 100644
--- a/src/module/AdiabaticCooling.cpp
+++ b/src/module/AdiabaticCooling.cpp
@@ -15,7 +15,7 @@ AdiabaticCooling::AdiabaticCooling(ref_ptr<AdvectionField> advectionField, doubl
 void AdiabaticCooling::process(Candidate *c) const {
 
 	Vector3d pos = c->current.getPosition();
-	double E = c->current.getEnergy(); // Note we use E=p/c (relativistic limit)
+	double E = c->current.getEnergy();
     double time = c->getTime();
 	
 	double Div = 0.;	
@@ -30,14 +30,17 @@ void AdiabaticCooling::process(Candidate *c) const {
 	double dEdt = -E / 3. * Div; 	// cooling due to advection -p/3 * div(V_wind)
 					// (see e.g. Kopp et al. Computer Physics Communication 183
 					// (2012) 530-542)
-	double dt = c->getCurrentStep() / c_light;
+	double v = c->current.getVelocityExact().getR();
+	if (v <= 0.)
+		return;
+	double dt = c->getCurrentStep() / v;
 	double dE = dEdt * dt;
 	
 	c->current.setEnergy(E + dE);
 	if (dEdt==0) {
 		return;
 	}	
-	c->limitNextStep(limit * E / fabs(dEdt) *c_light);
+	c->limitNextStep(limit * E / fabs(dEdt) * v);
 }
 
 void AdiabaticCooling::setLimit(double l) {
