diff -ruN /mnt/data/work_inc/CRPropa3-master/include/crpropa/module/PropagationBP.h patched/include/crpropa/module/PropagationBP.h
--- /mnt/data/work_inc/CRPropa3-master/include/crpropa/module/PropagationBP.h	2026-04-17 07:38:11.887302935 +0000
+++ patched/include/crpropa/module/PropagationBP.h	2026-04-17 08:31:12.722908744 +0000
@@ -113,7 +113,7 @@
 	 * @param q		 current charge of the candidate 
 	 * @param m		 current rest mass of the candidate
 	 */
-	void tryStep(const Y &y, Y &out, Y &error, double h, ParticleState &p, double z, double q, double m) const;
+	void tryStep(const Y &y, Y &out, Y &error, double dt, double step, ParticleState &p, double z, double q, double m) const;
 
 	/** Set functions for the parameters of the class PropagationBP */
 
diff -ruN /mnt/data/work_inc/CRPropa3-master/src/module/PropagationBP.cpp patched/src/module/PropagationBP.cpp
--- /mnt/data/work_inc/CRPropa3-master/src/module/PropagationBP.cpp	2026-04-17 07:38:24.391392141 +0000
+++ patched/src/module/PropagationBP.cpp	2026-04-17 08:31:12.723535759 +0000
@@ -5,14 +5,14 @@
 #include <vector>
 
 namespace crpropa {
-	void PropagationBP::tryStep(const Y &y, Y &out, Y &error, double h,
+	void PropagationBP::tryStep(const Y &y, Y &out, Y &error, double dt, double step,
 			ParticleState &particle, double z, double q, double m) const {
-		out = dY(y.x, y.u, h, z, q, m);  // 1 step with h
+		out = dY(y.x, y.u, dt, z, q, m);  // 1 step with dt
 
-		Y outHelp = dY(y.x, y.u, h/2, z, q, m);  // 2 steps with h/2
-		Y outCompare = dY(outHelp.x, outHelp.u, h/2, z, q, m);
+		Y outHelp = dY(y.x, y.u, dt/2, z, q, m);  // 2 steps with dt/2
+		Y outCompare = dY(outHelp.x, outHelp.u, dt/2, z, q, m);
 
-		error = errorEstimation(out.x , outCompare.x , h);
+		error = errorEstimation(out.x , outCompare.x , step);
 	}
 
 
@@ -113,7 +113,7 @@
 
 			// try performing step until the target error (tolerance) or the minimum/maximum step size has been reached
 			while (true) {
-				tryStep(yIn, yOut, yErr, step / std::max(current.getVelocityExact().getR(), 1e-300), current, z, q, m);
+				tryStep(yIn, yOut, yErr, step / std::max(current.getVelocityExact().getR(), 1e-300), step, current, z, q, m);
 				r = yErr.x.getR() / std::max(step, 1e-300) / tolerance;  // ratio of absolute direction error and tolerance
 				if (r > 1) {  // large direction error relative to tolerance, try to decrease step size
 					if (step == minStep)  // already minimum step size
